File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.ts
function searchRange(nums: number[], target: number): number[] {
    const answer:number[] = [-1,-1];
    answer[0] = searchPosition(nums, target, true);
    if(answer[0] !== -1){
        answer[1] = searchPosition(nums,target, false);
    }
    return answer;
};

const searchPosition = (nums:number[],target:number,isFirstIndex:boolean):number => {
    let [answer,start] = [-1,0];
    let end:number = nums.length - 1;
    
    while(start <= end){
        let middle:number = Math.floor(start + (end - start) / 2);
        if(target < nums[middle]){
            end = middle - 1;
        }else if(target > nums[middle]){
            start = middle + 1;
        }else{
            answer = middle;
            if(isFirstIndex){
                end = middle - 1;
            }else{
                start = middle + 1;
            }
        }
    }
    return answer;
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] answer = {-1,-1};
        
        answer[0] = searchPosition(nums, target, true);
        if(answer[0] != -1){
            answer[1] = searchPosition(nums, target, false);
        }
        return answer;
    }
    int searchPosition(int[] arr, int target, boolean isFirstIndex){
        int ans = -1;
        int start = 0;
        int end = arr.length - 1;
        
        while(start <= end){
            int middle = start + (end - start) / 2;
            
            if(target < arr[middle]){
                end = middle - 1;
            }
            else if(target > arr[middle]){
                start = middle + 1;
            }
            else{
                ans = middle;
                if(isFirstIndex){
                    end = middle - 1;
                }
                else{
                    start = middle + 1;
                }
            }
        }
        return ans;
    }
}